src/core/net/srp_server.cpp

```diff
@@ -0,0 +1,1604 @@
+                                                            uint16_t                aDnsMessageId)
+{
+    const UpdateMetadata *ret = nullptr;
+
+    for (const UpdateMetadata *update = mOutstandingUpdates.GetHead(); update != nullptr; update = update->GetNext())
+    {
+        if (aDnsMessageId == update->GetDnsHeader().GetMessageId() &&
+            aMessageInfo.GetPeerAddr() == update->GetMessageInfo().GetPeerAddr() &&
+            aMessageInfo.GetPeerPort() == update->GetMessageInfo().GetPeerPort())
+        {
+            ExitNow(ret = update);
+        }
+    }
+
+exit:
+    return ret;
+}
+
+void Server::HandleDnsUpdate(Message &                aMessage,
+                             const Ip6::MessageInfo & aMessageInfo,
+                             const Dns::UpdateHeader &aDnsHeader,
+                             uint16_t                 aOffset)
+{
+    otError   error = OT_ERROR_NONE;
+    Dns::Zone zone;
+    Host *    host = nullptr;
+
+    uint16_t headerOffset = aOffset - sizeof(aDnsHeader);
+
+    otLogInfoSrp("server: receives DNS update from %s", aMessageInfo.GetPeerAddr().ToString().AsCString());
+
+    SuccessOrExit(error = ProcessZoneSection(aMessage, aDnsHeader, aOffset, zone));
+
+    if (FindOutstandingUpdate(aMessageInfo, aDnsHeader.GetMessageId()) != nullptr)
+    {
+        otLogInfoSrp("server: drops duplicated SRP update request: messageId=%hu", aDnsHeader.GetMessageId());
+
+        // Silently drop duplicate requests.
+        // This could rarely happen, because the outstanding SRP update timer should
+        // be much shorter than the SRP update retransmission timer.
+        ExitNow(error = OT_ERROR_NONE);
+    }
+
+    // Per 2.3.2 of SRP draft 6, no prerequisites should be included in a SRP update.
+    VerifyOrExit(aDnsHeader.GetPrerequisiteRecordCount() == 0, error = OT_ERROR_FAILED);
+
+    host = Host::New(GetInstance());
+    VerifyOrExit(host != nullptr, error = OT_ERROR_NO_BUFS);
+    SuccessOrExit(error = ProcessUpdateSection(*host, aMessage, aDnsHeader, zone, headerOffset, aOffset));
+
+    // Parse lease time and validate signature.
+    SuccessOrExit(error = ProcessAdditionalSection(host, aMessage, aDnsHeader, headerOffset, aOffset));
+
+    HandleUpdate(aDnsHeader, host, aMessageInfo);
+
+exit:
+    if (error != OT_ERROR_NONE)
+    {
+        if (host != nullptr)
+        {
+            host->Free();
+        }
+        SendResponse(aDnsHeader, ErrorToDnsResponseCode(error), aMessageInfo);
+    }
+}
+
+otError Server::ProcessZoneSection(const Message &          aMessage,
+                                   const Dns::UpdateHeader &aDnsHeader,
+                                   uint16_t &               aOffset,
+                                   Dns::Zone &              aZone)
+{
+    otError   error = OT_ERROR_NONE;
+    Dns::Zone zone;
+
+    VerifyOrExit(aDnsHeader.GetZoneRecordCount() == 1, error = OT_ERROR_FAILED);
```

Comment:
```suggestion
    VerifyOrExit(aDnsHeader.GetZoneRecordCount() == 1, error = OT_ERROR_PARSE);
```
which DNS rcode is better? FORMAT_ERROR or REFUSED ?

Reply:
I think we should use FORMAT_ERROR. https://tools.ietf.org/html/rfc2136#section-3. Will update.